docs/eventing/sources/containersource/reference.md

```diff
@@ -0,0 +1,131 @@
+# ContainerSource reference
+
+This topic provides reference information about the configurable fields for the
+ContainerSource object.
+
+
+## ContainerSource
+
+A ContainerSource definition supports the following fields:
+
+| Field | Description | Required or optional |
+|-------|-------------|----------------------|
+| [`apiVersion`][kubernetes-overview] | Specifies the API version, for example `sources.knative.dev/v1`. | Required |
+| [`kind`][kubernetes-overview] | Identifies this resource object as a ContainerSource object. | Required |
+| [`metadata`][kubernetes-overview] | Specifies metadata that uniquely identifies the ContainerSource object. For example, a `name`. | Required |
+| [`spec`][kubernetes-overview] | Specifies the configuration information for this ContainerSource object. | Required |
+| [`spec.sink`](#sink-parameter) | A reference to an object that resolves to a URI to use as the sink. | Required |
+| [`spec.template`](#template-parameter) | A `template` in the shape of `Deployment.spec.template` to be used for this ContainerSource. | Required |
+| [`spec.ceOverrides`](#cloudevent-overrides) | Defines overrides to control the output format and modifications to the event sent to the sink. | Optional |
+
+
+### Sink parameter
+
+The `sink` parameter is a reference to an object that resolves to a URI to use as the sink.
+
+A `sink` definition supports the following fields:
+
+| Field | Description | Required or optional |
+|-------|-------------|----------------------|
+| `ref` | This points to an Addressable. | Required if _not_ using `uri`  |
+| `ref.apiVersion` | API version of the referent. | Required if using `ref` |
+| [`ref.kind`][kubernetes-kinds] | Kind of the referent. | Required if using `ref` |
+| [`ref.name`][kubernetes-names] | Name of the referent. | Required if using `ref` |
+| [`ref.namespace`][kubernetes-namespaces] | Namespace of the referent. If omitted this defaults to the object holding it. | Optional |
+| `uri` | This can be an absolute URL with a non-empty scheme and non-empty host that points to the target or a relative URI. Relative URIs are resolved using the base URI retrieved from Ref. | Required if _not_ using `ref` |
+
+!!! note
+    At least one of `ref` or `uri` is required. If both are specified, `uri` is
+    resolved into the URL from the Addressable `ref` result.
+
+#### Example: sink parameter
+
+Given the following YAML, if `ref` resolves into
+`"http://mysink.default.svc.cluster.local"`, then `uri` is added to this
+resulting in `"http://mysink.default.svc.cluster.local/extra/path"`.
+
+<!-- TODO we should have a page to point to describing the ref+uri destinations and the rules we use to resolve those and reuse the page. -->
+
+```yaml
```

Comment:
same as below re YAMLs, I think it would be better if we show example bits in context of a whole resource

Reply:
My aim with these snippets was to show an example using the fields for the specific parameter (eg the `sink` parameter). The full resource is shown on the procedure page, which links out to this page for more info about the options to use.

Would you be able to explain further how I can improve this or maybe give an example of what you mean? I am open to ideas :)

Reply:
I'd like for our snippets to be sort of like the example below, to show _where_ `sink` sits within the full YAML if that makes sense. Wrt to linking pages, I think we shouldn't assume that users will always see something that's on a different page, e.g. maybe they already know how to create a ContainerSource but they're just coming directly to this page as a quick reference to see where/how they'd add a sink in an existing YAML, and we don't really show them that if it's just a snippet without context.

I'd like to see this added to the style guide as well since there's at least one other open issue related to this https://github.com/knative/docs/issues/3703

```apiVersion: sources.knative.dev/v1
kind: ContainerSource
metadata:
  name: test-heartbeats
spec:
  ...
  sink:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: event-display

Reply:
I see what you mean. We do snippets like that in our docs but I was unsure whether to use that style in Knative. I'll update the snippets. :)

Reply:
Opened an issue to add this to the style guide also https://github.com/knative/docs/issues/3794